_posts/2024-10-24-autorebalancing-on-scaling.md

```diff
@@ -0,0 +1,152 @@
+---
+layout: post
+title:  "Auto-rebalancing on cluster scaling"
+date: 2024-10-24
+author: paolo_patierno
+---
+
+Scaling an Apache Kafka cluster up or down can help meet various goals and deliver multiple benefits.
+Scaling up by adding new brokers enables the cluster to handle a higher data load across clients.
+Scaling down by removing brokers can reduce costs and energy consumption when demand is lower.
+But scaling the cluster makes more sense if it's done automatically, for example by leveraging the HPA (Horizontal Pod Autoscaler) within Kubernetes.
+After scaling up, the issue arises from the cluster being unbalanced, with new brokers left empty.
+This requires rebalancing topic partitions to equally spread the load across all brokers, old and new.
+On the other hand, scaling down may not be possible if the brokers to be removed host topic partitions, unless we are willing to accept offline replicas and sacrifice High Availability (HA).
+In both scenarios, rebalancing the cluster by using the Cruise Control integration within Strimzi is the solution.
+Doing it automatically makes the process even easier and this is what we are going to talk about in this blog post.
+
+<!--more-->
+
+### How scaling and rebalancing is done today
+
+When scaling up a Kafka cluster, newly added brokers don't get any partitions for the topics hosted on the other brokers.
+The new brokers remain empty and will receive partitions for newly created topics based on Apache Kafka's distribution mechanism.
```

Comment:
Should we add a note box explaining how Kafka distribute partitions and replicas? Something like:

>[!NOTE]
> When you create a new topic, Kafka places the partitions and replicas in a way that the brokers with least number of existing partitions are used first, and replicas for same partition are on different brokers.

Reply:
Not sure we need this kind of details here tbh.